File: src/Haplotype.cpp
#include "../include/Haplotype.h"

using namespace std;

int Haplotype::score(const HaplotypePair& hap, const vector<int>& int_gts) {
	int	total_score = 0;
	for(size_t k = 0; k < int_gts.size(); ++k) {
		if(hap.first.hap[k] + hap.second.hap[k] == int_gts[k])
			total_score += 1;
	}
	return total_score;
}

int Haplotype::score(const Haplotype& hap_mat,
					const Haplotype& hap_pat, const vector<int>& int_gts) {
	int	total_score = 0;
	for(size_t k = 0; k < int_gts.size(); ++k) {
		if(hap_mat.hap[k] + hap_pat.hap[k] == int_gts[k])
			total_score += 1;
	}
	return total_score;
}

int Haplotype::match_score(const HaplotypePair& prev_hap,
										const HaplotypePair& hap) {
	const Haplotype&	prev_hap1 = prev_hap.first;
	const Haplotype&	prev_hap2 = prev_hap.second;
	const Haplotype&	hap1 = hap.first;
	const Haplotype&	hap2 = hap.second;
	return (prev_hap1.position == hap1.position ? 1 : 0) +
		   (prev_hap2.position == hap2.position ? 1 : 0);
}

vector<HaplotypePair> Haplotype::collect_optimal_haplotype_pairs(
											const vector<Haplotype>& haps_mat,
											const vector<Haplotype>& haps_pat,
											const vector<int>& int_gts) {
	int	max_score = 0;
	vector<HaplotypePair>	max_combs;
	for(auto p = haps_mat.begin(); p != haps_mat.end(); ++p) {
		const Haplotype&	hap_mat = *p;
		for(auto q = haps_pat.begin(); q != haps_pat.end(); ++q) {
			const Haplotype&	hap_pat = *q;
			const int	s = score(hap_mat, hap_pat, int_gts);
			if(s > max_score) {
				max_score = s;
				max_combs.clear();
				max_combs.push_back(HaplotypePair(hap_mat, hap_pat));
			}
			else if(s == max_score) {
				max_combs.push_back(HaplotypePair(hap_mat, hap_pat));
			}
		}
	}
	return max_combs;
}

vector<HaplotypePair> Haplotype::collect_max_score(
										const vector<HaplotypePair>& combs,
										const HaplotypePair& prev_hap) {
	int	max_score = 0;
	vector<HaplotypePair>	max_combs;
	for(auto p = combs.begin(); p != combs.end(); ++p) {
		const int	score = match_score(prev_hap, *p);
		if(score == max_score) {
			max_combs.push_back(*p);
		}
		else if(score > max_score) {
			max_score = score;
			max_combs.clear();
			max_combs.push_back(*p);
		}
	}
	return max_combs;
}

HaplotypePair Haplotype::impute(const vector<int>& int_gts,
								const vector<Haplotype>& haps_mat,
								const vector<Haplotype>& haps_pat,
								const HaplotypePair& prev_hap, int seed) {
	// blute-force
	const auto	combs = collect_optimal_haplotype_pairs(haps_mat,
														haps_pat, int_gts);
	
	// collect combinations that are most matched with the previous
	const auto	filtered_combs = collect_max_score(combs, prev_hap);
	
	// select a combination with random
	const int	j = seed % filtered_combs.size();
	return filtered_combs[j];
}

File: python/Haplotype.py
from __future__ import annotations

# coding: utf-8
# VCFImputable.py

from itertools import *
from abc import ABC, abstractmethod
from typing import Iterator, Tuple


#################### Haplotype ####################

class Haplotype(object):
	def __init__(self, h: list[int], sample_id: int, i: int):
		self.hap: list[int] = h
		self.position: tuple[int, int] = (sample_id, i)
	
	@staticmethod
	def default_value() -> Haplotype:
		hap: list[int] = []
		return Haplotype(hap, 0, 2)
	
	@staticmethod
	def score(int_gts: list[int], hap_mat: Haplotype,
								  hap_pat: Haplotype) -> int:
		return sum(1 for h1, h2, gt in zip(hap_mat.hap, hap_pat.hap, int_gts)
															if h1 + h2 == gt)
	
	@staticmethod
	def collect_optimal_haplotype_pairs(int_gts: list[int],
										haps_mat: list[Haplotype],
										haps_pat: list[Haplotype]
												) -> list[HaplotypePair]:
		scores = [ Haplotype.score(int_gts, hap_mat, hap_pat)
						for hap_mat, hap_pat in product(haps_mat, haps_pat) ]
		max_score = max(scores)
		return [ hap_pair
					for s, hap_pair in zip(scores, product(haps_mat, haps_pat))
															if s == max_score ]
	
	@staticmethod
	def match_score(prev_hap: HaplotypePair, hap: HaplotypePair) -> int:
		prev_hap1, prev_hap2 = prev_hap
		hap1, hap2 = hap
		return ((1 if prev_hap1.position == hap1.position else 0) +
				(1 if prev_hap2.position == hap2.position else 0))
	
	@staticmethod
	def collect_max_score(combs: list[HaplotypePair],
								prev_hap: HaplotypePair) -> list[HaplotypePair]:
		scores = [ (Haplotype.match_score(prev_hap, hap), hap)
													for hap in combs ]
		max_score = max(s for s, hap in scores)
		return [ hap for s, hap in scores if s == max_score ]
	
	@staticmethod
	def impute(int_gts: list[int], haps_mat: list[Haplotype],
						haps_pat: list[Haplotype], prev_hap: HaplotypePair,
						seed: int) -> HaplotypePair:
		# blute-force
		combs = Haplotype.collect_optimal_haplotype_pairs(int_gts,
														haps_mat, haps_pat)
		
		# collect combinations that are most matched with the previous
		filtered_combs = Haplotype.collect_max_score(combs, prev_hap)
		
		# select a combination with seed
		j = seed % len(filtered_combs)
		return filtered_combs[j]
	
HaplotypePair = Tuple[Haplotype, Haplotype]
